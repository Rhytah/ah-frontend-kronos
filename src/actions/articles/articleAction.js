import axios from 'axios';
import { toast } from 'react-toastify';
import { ArticleConstants } from '../actionTypes';
import { toastSuccess, toastFailure } from '../../utils/toast';
import history from '../../utils/history';

export function ArticleSuccess(successMessage) {
  return {
    type: ArticleConstants.ARTICLE_SUCCESS,
    payload: successMessage,
  };
}

export function ArticleFail(errorMessage) {
  return {
    type: ArticleConstants.ARTICLE_FAILURE,
    payload: errorMessage,
  };
}

export function ArticleRequest() {
  return {
    type: ArticleConstants.ARTICLE_REQUEST,
  };
}


export function ArticleDeleteSuccess(successMessage) {
  return {
    type: ArticleConstants.ARTICLE_DELETE_SUCCESS,
    payload: successMessage,
  };
}

export function ArticleDeleteFail(errorMessage) {
  return {
    type: ArticleConstants.ARTICLE_DELETE_FAILURE,
    payload: errorMessage,
  };
}

export function ArticleDeleteRequest() {
  return {
    type: ArticleConstants.ARTICLE_DELETE_REQUEST,
  };
}

const BASE_URL = 'https://ah-backend-kronos-staging.herokuapp.com/api/articles/';

export function saveArticle(newArticle) {
  // eslint-disable-next-line func-names
  return function (dispatch) {
    dispatch(ArticleRequest());
    return axios.post(BASE_URL, newArticle, {
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${sessionStorage.getItem('token')}`,
      },
    })
      .then((response) => {
        dispatch(ArticleSuccess(response.data));
        toast.dismiss();
        toastSuccess('Article created successfully', 'A');
        history.push(`/${response.data.article.slug}`);
      })
      .catch((error) => {
        toast.dismiss();
        toastFailure('Article creation failed! Authetication credentials were not provided', 'A');
        dispatch(ArticleFail(error.response.data));
      });
  };
}

export function updateArticle(newArticle, articleSlug) {
  // eslint-disable-next-line func-names
  return function (dispatch) {
    dispatch(ArticleRequest());
    return axios.patch(`${BASE_URL}${articleSlug}`, newArticle, {
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${sessionStorage.getItem('token')}`,
      },
    })
      .then((response) => {
        dispatch(ArticleSuccess(response.data));
        toast.dismiss();
        toastSuccess('Article successfully updated', 'A');
        history.push(`/${articleSlug}`);
      })
      .catch((error) => {
        toast.dismiss();
        toastFailure('Article update failed! Try agin later.', 'A');
        dispatch(ArticleFail(error.response.data));
      });
  };
}

export function deleteArticle(articleSlug) {
  // eslint-disable-next-line func-names
  return function (dispatch) {
    dispatch(ArticleDeleteRequest());
    return axios.delete(`${BASE_URL}${articleSlug}`, {
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${sessionStorage.getItem('token')}`,
      },
    })
      .then((response) => {
        dispatch(ArticleDeleteSuccess(response.data));
        toast.dismiss();
        toastSuccess('Article deleted successfully.', 'A');
        history.push('/articles');
      })
      .catch((error) => {
        toast.dismiss();
        toastFailure('Article delete failed! Try agin later.', 'A');
        dispatch(ArticleDeleteFail(error.response.data));
      });
  };
}
