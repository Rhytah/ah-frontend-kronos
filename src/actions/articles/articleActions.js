import axios from 'axios';
import { ArticleConstants } from '../actionTypes';

const getSingleArticle = slug => (dispatch) => {
  dispatch({
    type: ArticleConstants.FETCHING_ARTICLE,
  });
  return axios
    .get(`https://ah-backend-kronos-staging.herokuapp.com/api/articles/${slug}`)
    .then((res) => {
      dispatch({
        type: ArticleConstants.FETCH_AN_ARTICLE,
        payload: res.data.article,
      });
    })
    .catch((error) => {
      dispatch({
        type: ArticleConstants.ARTICLE_FETCH_NOT_FOUND,
        payload: error.response.data,
      });
    });
};
export default getSingleArticle;
