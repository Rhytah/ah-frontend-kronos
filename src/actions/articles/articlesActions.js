import axios from 'axios';
import { ArticleConstants } from '../actionTypes';

const fetchArticlesUrl = 'https://ah-backend-kronos-staging.herokuapp.com/api/articles/';
const fetchArticles = () => dispatch => axios.get(fetchArticlesUrl).then((response) => {
  dispatch({
    type: ArticleConstants.FETCH_ARTICLES,
    payload: response.data.article.results,
  });
});
export default fetchArticles;
