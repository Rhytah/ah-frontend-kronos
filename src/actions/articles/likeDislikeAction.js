import axios from 'axios';
import { ArticleConstants } from '../actionTypes';
import URLS from '../../urlConnections';

export const ArticleLikeSuccess = successMessage => ({
  type: ArticleConstants.ARTICLE_LIKING_SUCCESS,
  payload: successMessage,
});

export const ArticleLikeFail = errorMessage => ({
  type: ArticleConstants.ARTICLE_LIKING_FAILURE,
  payload: errorMessage,
});

export const ArticleDislikeSuccess = successMessage => ({
  type: ArticleConstants.ARTICLE_DISLIKING_SUCCESS,
  payload: successMessage,
});

export function ArticleDislikeFail(errorMessage) {
  return {
    type: ArticleConstants.ARTICLE_DISLIKING_FAILURE,
    payload: errorMessage,
  };
}

export const likingArticle = (articleSlug, articleLike) => dispatch => axios.put(`${URLS.BACKEND_URL}/api/articles/${articleSlug}/likes/`,
  articleLike, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${sessionStorage.getItem('token')}`,
    },
  })
  .then((response) => {
    const { like_status: { message } } = response.data;
    if (message === 'You have already liked the article') {
      dispatch(ArticleLikeFail(message));
    } else if (message === 'You have liked an article') {
      dispatch(ArticleLikeSuccess(response.data));
    } else if (message === 'You have already disliked the article') {
      dispatch(ArticleDislikeFail(message));
    } else if (message === 'You have disliked an article') {
      dispatch(ArticleDislikeSuccess(response.data));
    }
  })
  .catch((error) => {
    const { like_status: { detail } } = error.response.data;
    dispatch(ArticleLikeFail(detail));
  });
