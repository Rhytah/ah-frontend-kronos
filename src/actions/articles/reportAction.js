/* eslint-disable prefer-destructuring */
/* eslint-disable func-names */
import axios from 'axios';
import { toastSuccess, toastFailure } from '../../utils/toast';
import { ArticleConstants } from '../actionTypes';

const apiUrl = 'https://ah-backend-kronos-staging.herokuapp.com/api/articles/';
const lastPart = '/report/';

const reportArticleAction = (articleSlug, report) => dispatch => axios.post(`${apiUrl}${articleSlug}${lastPart}`, report, {
  headers: {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${sessionStorage.getItem('token')}`,
  },
})
  .then((response) => {
    toastSuccess(response.data.msg, 'A');
    dispatch({
      type: ArticleConstants.ARTICLE_REPORT_SUCCESS,
    });
  })
  .catch((error) => {
    const responseData = error.response;
    let errorMessage;
    switch (responseData.status) {
      case 422:
        errorMessage = responseData.data.message;
        break;
      default:
        [errorMessage] = responseData.data.errors.reason;
    }
    toastFailure(errorMessage, 'A');
    dispatch({
      type: ArticleConstants.ARTICLE_REPORT_FAIL,
    });
  });

export default reportArticleAction;
