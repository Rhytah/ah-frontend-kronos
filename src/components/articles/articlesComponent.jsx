/* eslint-disable react/no-unused-state */
import React, { Component } from 'react';
import { Link } from 'react-router-dom';

import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import fetchArticles from '../../actions/articles/articlesActions';
import '../../assets/scss/Landing.scss';

export class Articles extends Component {
  componentDidMount() {
    const { fetchArticles } = this.props;
    fetchArticles();
  }

  render() {
    const { articles } = this.props;
    console.log('component articles', articles);

    const articlesList = articles.map(article => (
      <div className="col-sm-6 item" key={article.slug}>
        <div className="row">
          <div className="col-md-12 col-lg-5">
            <Link to={`/${article.slug}`}>
              <img
                className="img-fluid"
                alt="article_image"
                src="https://cdn.pixabay.com/photo/2019/07/07/14/03/fiat-4322521__480.jpg"
              />
            </Link>
          </div>
          <div className="col">
            <Link to={`/${article.slug}`}>
              <h4 className="name">{article.title}</h4>
            </Link>
            <p className="description">{article.description}</p>
            <div className="article-details">
              <p>
                By:
                {article.author.username}
              </p>
              <span>
                {article.read_time}
                {' '}
Read
              </span>
              <i className="far fa-thumbs-up">45</i>
              <span />
              <i className="far fa-thumbs-down">12</i>
              <br />
            </div>
          </div>
        </div>
      </div>
    ));
    return (
      <div>
        <div className="projects-horizontal">
          <div className="container">
            <div className="intro">
              <h2 className="text-center">Articles</h2>
              <p className="text-center">Browse all articles</p>
            </div>
            <form className="search-form">
              <div className="input-group">
                <input className="form-control" type="text" placeholder="I am looking for.." />
                <div className="input-group-append">
                  <button className="btn btn-light" type="button">
                    Search
                    {' '}
                  </button>
                </div>
              </div>
            </form>

            <div className="row projects">{articlesList}</div>
          </div>
        </div>
      </div>
    );
  }
}
Articles.propTypes = {
  fetchArticles: PropTypes.func.isRequired,
  articles: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string,
      description: PropTypes.string,
      slug: PropTypes.string,
      body: PropTypes.string,
      read_time: PropTypes.string,
      author: PropTypes.shape({
        username: PropTypes.string,
        bio: PropTypes.string,

      }),
    }),
  ).isRequired,
};
// Articles.defaultProps = {
//   articles: [],
// };
export const mapStateToProps = state => ({ articles: state.articlesReducer.articles });

export default connect(
  mapStateToProps,
  { fetchArticles },
)(Articles);
