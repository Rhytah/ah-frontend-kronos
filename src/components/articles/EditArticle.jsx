/* eslint-disable jsx-a11y/label-has-for */
/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import Input from '../common/Input';
import Button from '../common/Button';
import TextField from '../common/TextField';
import '../../assets/css/Article.css';
import * as articleActions from '../../actions/articles/articleAction';

export class EditArticle extends Component {
  constructor(props) {
    super(props);

    this.state = {
      title: '',
      description: '',
      body: '',
      image: '',
      slug: '',
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount() {
    // eslint-disable-next-line react/destructuring-assignment
    const { articleData } = this.props.location.state;
    this.setState({
      title: articleData.title,
      description: articleData.description,
      body: articleData.body,
      slug: articleData.slug,
      image: articleData.image,
    });
  }

  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  };

  handleSubmit = (event) => {
    event.preventDefault();
    const {
      title, description, body, image, slug,
    } = this.state;
    const { actions } = this.props;
    const newarticle = {
      title,
      description,
      body,
      image,
    };
    actions.updateArticle(newarticle, slug);
    this.handleClearForm();
  };

  handleClearForm() {
    this.setState({
      title: '',
      description: '',
      body: '',
      image: '',
    });
  }

  render() {
    const {
      title, description, body, image, slug,
    } = this.state;
    return (
      <section className="create-article-section">
        <h3 className="text-center">Edit an article</h3>
        <form onSubmit={this.handleSubmit}>
          <label>Title</label>
          <Input
            type="text"
            name="title"
            value={title}
            placeholder="Type article title here...."
            className="form-control"
            handleChange={this.handleChange}
            required={false}
          />
          <label>Slug</label>
          <Input
            type="text"
            name="slug"
            value={slug}
            placeholder="Type article slug here...."
            className="form-control"
            handleChange={this.handleChange}
          />
          <label>Image URL</label>
          <Input
            type="url"
            name="image"
            value={image}
            placeholder="https://example.com"
            className="form-control"
            handleChange={this.handleChange}
            required={false}
          />
          <label>Description</label>
          <TextField
            rows="3"
            name="description"
            value={description}
            placeholder="Type article description here..."
            className="form-control"
            handleChange={this.handleChange}
            required={false}
          />
          <label>Body</label>
          <TextField
            rows="17"
            name="body"
            value={body}
            placeholder="Type article body here..."
            className="form-control"
            handleChange={this.handleChange}
            required={false}
          />
          <div className="form-group">
            <Link to={slug} className="btn btn-danger">
              Cancel
            </Link>
            <Button className="btn btn-primary" type="submit">
              Update Article
            </Button>
          </div>
        </form>
      </section>
    );
  }
}

EditArticle.defaultProps = {
  location: {},
};

EditArticle.propTypes = {
  actions: PropTypes.shape({ registerArticle: PropTypes.func }).isRequired,
  location: PropTypes.shape(),
};

export const mapStateToProps = state => ({
  article: state.article,
});

export const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(articleActions, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(EditArticle);
