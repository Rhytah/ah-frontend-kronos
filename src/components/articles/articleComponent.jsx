import React, { Component } from 'react';
import '../../assets/scss/article.scss';
import { Link } from 'react-router-dom';

import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import getSingleArticle from '../../actions/articles/articleActions';
import Loading from '../common/Loading';
import { deleteArticle } from '../../actions/articles/articleAction';

export class Article extends Component {
  constructor(props) {
    super(props);
    this.state = {
      slug: '',
    };

    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount() {
    // eslint-disable-next-line no-shadow
    const { getSingleArticle } = this.props;
    const {
      match: {
        params: { slug },
      },
    } = this.props;
    getSingleArticle(slug);
    this.setState({
      slug,
    });
  }

  handleSubmit() {
    const { slug } = this.state;
    // eslint-disable-next-line no-shadow
    const { deleteArticle } = this.props;
    deleteArticle(slug);
  }


  render() {
    const { article, isFetching } = this.props;

    const createdAt = new Date(article.created_at).toLocaleDateString('en-GB', {
      day: 'numeric',
      month: 'short',
      year: 'numeric',
    });
    if (isFetching) {
      return <Loading className="article-loader" />;
    }
    return (
      <div className="article-dual-column">
        <div className="container">
          <div className="row row-img">
            <div className="col-md-10 offset-md-1">
              <div className="intro">
                <h1 className="text-center">{article.title}</h1>
              </div>
            </div>
          </div>
          <section>
            <div className="row article-row">
              <div className="col-md-10 col-lg-7 offset-md-1 offset-lg-0">
                <div className="article-details">
                  <span />
                  <span>{article.author.username}</span>
                  <span className="date">{createdAt}</span>

                  <span>
                    {article.read_time}
                    {' '}
read
                  </span>
                </div>

                <div className="text">
                  <img alt="" src={article.image} className="img-fluid" />

                  <p>{article.body}</p>
                </div>
                <p className="article-icons">
                  <i className="far fa-bookmark" />
                  <i className="fa fa-heart" />
                  <i className="far fa-thumbs-up">{article.likes_count}</i>
                  { sessionStorage.getItem('username') === article.author.username
                    ? (
                      <React.Fragment>
                        <Link
                          to={{
                            pathname: '/update-article',
                            state: {
                              articleData: {
                                title: article.title,
                                description: article.description,
                                body: article.body,
                                slug: article.slug,
                                image: article.image,
                              },
                            },
                          }}
                          className="fa fa-edit"
                        />
                        <Link
                          to={{
                            state: {
                              articleData: {
                                title: article.title,
                                description: article.description,
                                body: article.body,
                                slug: article.slug,
                                image: article.image,
                              },
                            },
                          }}
                          className="fa fa-trash"
                          data-toggle="modal"
                          data-target="#exampleModal"
                        />
                      </React.Fragment>
                    ) : null
                }
                </p>
                <ul className="tags">
                  <li className="tag">Tag1</li>
                  <li className="tag">Tag2</li>
                  <li className="tag">Tag3</li>
                </ul>
                <div
                  className="modal fade"
                  id="exampleModal"
                  tabIndex="-1"
                  role="dialog"
                  aria-labelledby="exampleModalLabel"
                  aria-hidden="true"
                >
                  <div className="modal-dialog" role="document">
                    <div className="modal-content">
                      <div className="modal-header">
                        <h5 className="modal-title" id="exampleModalLabel">
                          {article.title}
                        </h5>
                        <button
                          type="button"
                          className="close"
                          data-dismiss="modal"
                          aria-label="Close"
                        >
                          <span aria-hidden="true">&times;</span>
                        </button>
                      </div>
                      <div className="modal-body">
                        Are you sure you want to delete this article?
                      </div>
                      <div className="modal-footer">
                        <button type="button" className="btn btn-secondary" data-dismiss="modal">
                          Cancel
                        </button>
                        <button
                          type="submit"
                          onClick={this.handleSubmit}
                          className="btn btn-danger"
                          data-dismiss="modal"
                        >
                          Yes
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </section>
        </div>
      </div>
    );
  }
}
Article.propTypes = {
  getSingleArticle: PropTypes.func.isRequired,
  article: PropTypes.shape({}),
  isFetching: PropTypes.bool.isRequired,
  deleteArticle: PropTypes.func,
  match: PropTypes.shape({
    params: PropTypes.shape({
      slug: PropTypes.string.isRequired,
    }),
  }).isRequired,
};
Article.defaultProps = {
  article: {},
  deleteArticle: () => {},
};
export const mapStateToProps = (state) => {
  const { article, isFetching } = state.singleArticleReducer;
  return { article, isFetching };
};

export default connect(
  mapStateToProps,
  { getSingleArticle, deleteArticle },
)(Article);
