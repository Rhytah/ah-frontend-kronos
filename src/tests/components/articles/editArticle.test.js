import React from 'react';
import { shallow } from 'enzyme';
import { EditArticle, mapStateToProps, mapDispatchToProps } from '../../../components/articles/EditArticle';

function renderEditArticleForm(args) {
  const defaultProps = {
    actions: { updateArticle: jest.fn() },
    article: {},
    location: {
      state: {
        articleData: {},
      },
    },
  };
  const props = { ...defaultProps, ...args };
  return shallow(<EditArticle {...props} />);
}

describe('create article page', () => {
  const wrapper = renderEditArticleForm();
  const wrapperInst = wrapper.instance();


  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle onChange event', () => {
    wrapper.setState({
      target: {
        name: 'title',
        value: 'new title',
      },
      preventDefault: jest.fn(),
    });
    const wrapperInstance = wrapper.instance();
    const event = {
      target: {
        name: 'title',
        value: 'new title 2',
      },
    };
    jest.spyOn(wrapperInstance, 'handleChange');
    wrapperInstance.handleChange(event);

    expect(wrapperInstance.handleChange).toBeCalled();
  });

  it('should handle submit and clear form after', () => {
    wrapper.setState({
      title: 'article title',
      body: 'article body',
      description: 'article description',
      image: 'https://image.com',
    });

    expect(wrapperInst.state.title).toBe('article title');
    expect(wrapperInst.state.body).toBe('article body');
    expect(wrapperInst.state.description).toBe('article description');
    expect(wrapperInst.state.image).toBe('https://image.com');

    const event = {
      target: {
        type: 'submit',
      },
      preventDefault: jest.fn(),
    };
    wrapperInst.handleSubmit(event);
    expect(wrapperInst.props.actions.updateArticle).toBeCalled();
    expect(wrapperInst.state.title).toBe('');
    expect(wrapperInst.state.body).toBe('');
    expect(wrapperInst.state.description).toBe('');
    expect(wrapperInst.state.image).toBe('');
  });

  it('should map tostate to props', () => {
    const mockedState = {
      article: {
        saving: true,
      },
    };
    const state = mapStateToProps(mockedState);
    expect(state.article.saving).toBe(true);
  });

  it('should map dispatch to the articleSuccess action', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).actions.ArticleSuccess();
    expect(dispatch.mock.calls[0][0]).toEqual({ type: 'SAVE_ARTICLE_SUCCESS' });
  });
});
