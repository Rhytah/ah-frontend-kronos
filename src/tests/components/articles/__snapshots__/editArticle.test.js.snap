// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create article page should match the snapshot 1`] = `ShallowWrapper {}`;
